### DISCUSSION TITLE
SpatieMediaLibrary - for each file, create model record

### DISCUSSION THREAD

**Discussion by diogogpinto** [filamentphp/filament#8579] [Help]
Hello everyone,

I'm using SpatieMediaLibrary in a RelationshipManager. In the form I have this:

```php
                Forms\Components\SpatieMediaLibraryFileUpload::make('documents')
                    ->label(__('File(s)'))
                    ->columnSpanFull()
                    ->multiple()
                    ->openable(),
```

I want to be able to select multiple files and, with every file, create a different record. I thought of editing the CreateAction and when I'm creating - loop through the files ($data) and create one record for each. So I've done this:

```php
                Tables\Actions\CreateAction::make()
                    ->using(function (array $data) {
                        dd($data); //for testing
                    }),
```

**Problem is**: using SpatieMediaLibraryFileUpload, the files don't go through the array $data, so I can't loop through them.

I'm not sure if I'm being clear, but I wonder if anyone can help?

**Comment by leandrocfe** [comment]
`mountedTableActionsData`?

```php
Tables\Actions\CreateAction::make()
    ->action(function () {
        //$this->mountedTableActionsData;
    })
```

**Reply by diogogpinto** [reply]
Hello @leandrocfe 

It returns NULL... I can't seem to get the Upload Files to loop through them. I'm in a loop myself because I can't seem to make this work...

```php
            Actions\Action::make('uploadDocuments')
                ->form([
                    Forms\Components\SpatieMediaLibraryFileUpload::make('documents')
                        ->collection('user-documents')
                        ->multiple(),
                ])
                ->action(function () {
                    dd($this->mountedActionsData);
                }),
```

**Reply by leandrocfe** [reply]
what about using [FileUpload](https://filamentphp.com/docs/3.x/forms/fields/file-upload)?
```php
->action(function (array $data) {
    dd($data);
})
```

**Reply by diogogpinto** [reply]
Thanks so much for your input @leandrocfe

That would pass the file name - but I'm having trouble accessing the Temporary Files and move them to the specified disk with Spatie's MediaLibrary. I'm looking under the hood, but not having much success.

**Reply by underdpt** [reply]
I think you can use `mutateDehydratedStateUsing` on your form input, in where you can get the `$state` which will be an array of `TemporaryUploadedFile`.

You can process your files there.

**Reply by diogogpinto** [reply]
@underdpt thank you for your input!

It returns an array:

```
array:2 [▼ // app/Filament/Resources/UserResource/Pages/ViewUser.php:48
  0 => "7.png"
  1 => "UnidAloj-011668.pdf"
]
```

But I can't seem to get to the temporary files. I also tried with saveUploadedFileUsing()

**Reply by underdpt** [reply]
Can you swap SpatieMediaLibraryFileUpload with a vanilla FileUpload?

**Reply by diogogpinto** [reply]
@underdpt I think I got it!

with saveUploadedFileUsing(), the $state returns an array of TemporaryUploadedFiles.

So, this is actually working:
```php
->form([
    Forms\Components\FileUpload::make('documents')
        ->preserveFilenames()
        ->multiple()
        ->saveUploadedFileUsing(function ($state) {
            $files = [];
            foreach ($state as $file) {
                $files[] = $file->getPath() . '/' . $file->getFilename();
            }
            dd($files);
        }),
]),
```

Thank you so much! Now I can go into each file and associate it with the model using the medialibrary methods.

**Reply by leandrocfe** [reply]
@diogopintogeridoc can you try?
```php
Action::make('uploadDocuments')
    ->form([
        Forms\Components\SpatieMediaLibraryFileUpload::make('documents')
            ->dehydrated(true)
            ->multiple(),
    ])
    ->action(function (array $data) {
        dd($data);
    })
```

**Reply by z4yed** [reply]
@dpgdoc just [prevent](https://filamentphp.com/docs/3.x/forms/fields/file-upload#preventing-files-from-being-stored-permanently) the file getting stored initially. Then you'll get a TemporaryUploadedFile instance.

```php
return $form
    ->schema([
        FileUpload::make('attachments')
            ->multiple()
            ->storeFiles(false)
            ->required(),
    ])
```
